pkg/stanza/fileconsumer/trie.go

```diff
@@ -0,0 +1,90 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+// TRIE data structure inspired by https://github.com/dghubble/trie
+/*
+	This differs from the original trie.
+   	This has been modified to detect partial matches as well.
+   	For eg.
+		If we add "ABCDEF" to this trie, and try to check if "ABCD" is in the trie,
+		it will return true because that's how fingerprint matching works in current implementation.
+*/
+
+package fileconsumer // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer"
+
+type Trie struct {
+	_        interface{}
+	children map[byte]*Trie
+}
+
+// NewPathTrie allocates and returns a new *Trie.
+func NewTrie() *Trie {
+	return &Trie{}
+}
+
+func (trie *Trie) HasKey(key []byte) bool {
+	node := trie
+	for _, r := range key {
+		node = node.children[r]
+		if node == nil {
+			return false
+		}
+		// We have reached end of the current path and all the previous characters have matched
+		// Return if current node is leaf and it is not root
+		if node.isLeaf() && node != trie {
+			return true
+		}
+	}
+	return false //  If it's an exact match, the final node will be a leaf.
+}
+
+// Put inserts the key into the trie
+func (trie *Trie) Put(key []byte) {
+	node := trie
+	for _, r := range key {
+		child, ok := node.children[r]
+		if !ok {
+			if node.children == nil {
+				node.children = map[byte]*Trie{}
+			}
+			child = NewTrie()
+			node.children[r] = child
+		}
+		node = child
+	}
+}
+
+// Delete removes the value associated with the given key. Returns true if a
+// node was found for the given key. If the node or any of its ancestors
+// becomes childless as a result, it is removed from the trie.
+func (trie *Trie) Delete(key []byte) bool {
```

Comment:
When is it intended that we will call `Delete`? Specifically I'm trying to understand how it relates to the lifecycle of a given file.

Reply:
We call `delete`, only after the file is successfully read and all the logs are emitted and it is added to the list of knownReaders.

Reply:
Here's a case that doesn't make sense to me with this design:

- First poll finds a file w/ "ABCDEF", add it to the trie, and start reading.
- Second poll finds "ABCDEF" unchanged, but also finds another file w/ "ABC".  `HasKey("ABC")` tells us that this file is not currently being read, so we call `Put("ABC")` but this has no effect on the trie because "ABCDEF" is already there.
- Then "ABCDEF" is fully read, so we add it to `knownReaders` and call `Delete("ABCDEF")`. This effectively deletes both "ABCDEF" and "ABC".
- A third poll now finds "ABC" again and `HasKey("ABC")` tells us it is not currently being read, so we start reading it again. Now we have two active readers for the same file resulting in duplicated logs.

This seems to indicate that we need to make use of values on nodes:
- `Put("ABC")` should store a value on "C"
- `HasKey("ABC")` should return true if "C" exists and has a value (regardless of whether it is a leaf)

However, this still may not be enough. Another scenario:
- First poll finds both "ABC" and "ABCDEF".
- The order in which we process these files changes the result:
  - `HasKey("ABC") => false` => `Put("ABC")` => `HasKey("ABCDEF") => true`
    - Results in trie that contains "ABC" only
  - `HasKey("ABCDEF") => false` => `Put("ABCDEF")` => `HasKey("ABC") => false` => `Put("ABC")`
    - Results in trie that contains "ABC" and "ABCDEF"

I think the correct behavior depends on the context but it's not clear to me how best to model this. If we find "ABC" in the first poll and then "ABCDEF" in the next, the trie should have only "ABCDEF". However, if find them in the opposite order or within the same poll, I think we should treat these as separate files. Do you have any ideas how to address this?

Reply:
Ok this one is a unique situation.
I think we should discuss how current file reading works.
So what happens in current implementation:

- It will loop through all the file paths.
- It will create fingerprints, then check for duplicates.

While checking duplicates, it will check both the ways.
Basically it will check if "ABCDEF" and "ABCD" has a common prefix, and if that's true, it will treat them as duplicates.
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/stanza/fileconsumer/file.go#L228

Reply:
Now to handle your scenario, we can use values in trie or we can go as per current implementation.

Reply:
@djaglowski idk but I feel like we should go as per current implementation. I agree it makes little sense to match "ABC" after adding "ABCDE", but that's how it works. I can also go the other way i.e. to introduce value.

Reply:
I believe the current behavior is unnecessarily aggressive in deduplicating files. It is appropriate to use the "StartsWith" logic when comparing files at subsequent times, but we have no explicit use case for using "StartsWith" when comparing files found _at the same time_. As a result, I've proposed https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/24235 to fix the problem.

Reply:
Ok, so what I think we can do here is introduce values. We can use a simple bool indicating that the current node is an end of the fingerprint.
We add `ABCDE` in the first poll:
`A - B - C - D - E (true)` we store the value i.e. true in the final node.
We then go ahead and add `ABC` in the second poll and the trie becomes:
`A - B - C (true) - D - E (true)` we store the value i.e. true in the final node.
When deleting. let's consider two scenarios:

- Delete `ABC` first: We remove the value from C, check for any child nodes, and exit since D is a child of C.
   Trie becomes, `A - B - C - D - E (true)`
- Delete `ABCDE` first: We remove the value from E and check for any child nodes. 
   We also check if we have stored a value in any other nodes in the path, if there are any such nodes, we simply exit at this stage.
   We would delete `E` and `D, then for `C` we would stop as there is a value stored here.
   Trie becomes `A - B - C (true)`

Reply:
@djaglowski have a look at the latest changes and test cases. I think it covers the scenarios.

Reply:
@VihasMakwana, the `HasKey` function wasn't updated to account for values on nodes. It still returns true only if the node is a leaf, so `HasKey(ABC)` would return false in the scenario where both `ABC` and `ABCDE` have been `Put` into the trie.

More importantly, I think the design you described introduces new problems.

1. Let's say we've added `ABC` and `ABCDE` to the trie, so we have `A - B - C (true) - D - E (true)`. Next we find two files: `ABCDEXYZ` and `ABCXYZ`. Are we able to recognize that these are the same two files we previously added? `HasKey(ABCDEXYZ)` can return `true` based on reaching a leaf node, but how does `HasKey(ABCXYZ)` return `true`? I think we need an additional case that returns true, roughly `if C.hasValue && !C.hasChild(X)`.
2. If we `Put(ABC)` and then in a subsequent poll find `ABCXYZ`, we can recognize that this is the same file with some new information, so if we `Put(ABCXYZ)`, we should end up with `A - B - C - X - Y - Z (true)`. Currently, we would add a second entry to the trie, indicating that we are reading two different files. How do we reconcile this? 
3. (A case to consider when solving scenario 2) Let's again say we have a trie `A - B - C (true) - D - E (true)`. Next we find `ABCDEXYZ` and ``ABCXYZ`. What _should_ happen, is that we recognize both of these are files are in the trie and then we update both to result in:
 ```
               /  X - Y - Z (true)
A - B - C 
               \  D - E - X - Y - Z (true)
```
The challenge here is that we need a rule regarding which value is "pushed" further down the trie when we update a key. Starting from `A - B - C (true) - D - E (true)`, if we only `Put(ABCDEXYZ`, we could end up with either `A - B - C (true) - D - E - X - Y - Z (true)` or `A - B - C - D - E (true) - X - Y - Z (true)`. We need to be sure that it is `A - B - C (true) - D - E - X - Y - Z (true)` so that we can `Put(ABCXYZ)` and "push" the value at C down. If instead we would up with `A - B - C - D - E (true) - X - Y - Z (true)`, then `Put(ABCXYZ)` would result in a trie with three keys. I think the solution here relies on roughly the same check mentioned above. When deciding if a value should be pushed, we should choose _not_ to push it `if C.hasValue && C.hasChild(D)`.

Reply:
Okay, got this stuff. 
For updating the value part, I think we should cover that in threadpool PR

Reply:
Okay, is forgot to commit the HasKey part. I have updated it. and added some more test cases

Reply:
> Okay, got this stuff. For updating the value part, I think we should cover that in threadpool PR

I think we should merge the trie only when it represents what we agree is the correct design for the problem.

Reply:
Makes sense, will go through the updating logic in PUT

Reply:
I have one scenario to confirm:
if we have `A - B - C (true) - D - E (true) - F - X -Y (true)`

- if we try to add, `A - B - C - D - X` to the above trie, we shouldn't add a duplicate entry but push the value at `C` down. Trie would become 
```
A - B - C - D - E (true) - F - X -Y (true)
            |
           X (true)
```
- if we try to add `A - B - C - D`  to the above trie, we shouldn't add a duplicate entry but push the value at `C` down. Trie would become `A - B - C - D (true) - E (true) - F - X -Y (true)`
- if we try to add, `A - B - C - D - E - F - X - Y- Z` to the above trie, we shouldn't add a duplicate entry, but push the value at `E` down. Trie would become `A - B - C (true) - D - E - F - X  - Y - Z(true)`

Reply:
In other words, we are pushing the last value encountered while traversing the current path?

Reply:
I _think_ that makes sense. At least I can't think of a scenario where that ends up giving us the wrong result.

Reply:
Okay, I pushed the changes.

Reply:
@djaglowski